sprint_16

KatSchreiner opened the pull request.

KatSchreiner committed:
Реализован экран онбординга с использованием UIPageViewController и UIPageControl

KatSchreiner committed:
Создан экран со списком категорий
Реализована логика создания новой категории
Экран со списком категорий переписан на архитектуру MVVM

Andreynnt commented on Tracker/CategoryViewController/CategoriesViewModel.swift:
```diff
@@ -0,0 +1,38 @@
+    var onCategoryAdded: (() -> Void)?
+    
+    var numberOfCategories: Int {
+        return categories.count
```
В однострочных функциях/вычисляемых свойствах return можно не писать)

Andreynnt commented on Tracker/CategoryViewController/CategoriesViewModel.swift:
```diff
@@ -0,0 +1,38 @@
+    func loadCategories() {
+        let categoriesList = categoryStore.categories.map { $0.title }
+        categories = categoriesList.filter { !$0.isEmpty }
+        onCategoriesChanged?()
```
Лучше сделать binding и привязать вызов метода в didSet к свойству categories

Andreynnt commented on Tracker/CategoryViewController/CategoriesViewController.swift:
```diff
@@ -0,0 +1,200 @@
+        super.viewDidLoad()
+        setupView()
+        setupBindings()
+        viewModel.loadCategories()
```
ViewModel сама должна решать, что ей делать - лучше назвать метод viewModel.viewDidLoad() и в нем уже  viewModel загрузит данные. ViewController должен быть максимально "Тупым" и просто уведомлять о своих событиях

Andreynnt commented on Tracker/CategoryViewController/CategoriesViewController.swift:
```diff
@@ -0,0 +1,200 @@
+            self?.showPlaceholder()
+        }
+
+        viewModel.onCategoryAdded = { [weak self] in
```
Можно удалить этот биндинг и оставить только onCategoriesChanged, потому что они одинаковые и по смыслу `onCategoryAdded` входит в `onCategoriesChanged`

Andreynnt commented on Tracker/CategoryViewController/CategoriesViewController.swift:
```diff
@@ -0,0 +1,200 @@
+        setupView()
+        setupBindings()
+        viewModel.loadCategories()
+        showPlaceholder()
```
Это тоже должна решать viewModel по-хорошему. И говорить viewController-y что нужно скрыть или показать placeholder

Andreynnt commented on Tracker/Onboarding/SinglePageViewController.swift:
```diff
@@ -0,0 +1,58 @@
+
+final class SinglePageViewController: UIViewController {
+    
+    lazy var backgroundImage: UIImageView = {
```
Не забывай про private у свойств

KatSchreiner committed:
Ревью. Исправила критические замечания:
1) удалила onCategoryAdded
2) viewModel управляет placeholder
3) во viewModel добавила viewDidLoad() и в нем загружаю данные
4) к свойству categories в didSet привязала вызов метода onCategoriesChanged
5) добавила private у свойств в SinglePageViewController
